R-package/R/lgb.DataProcessor.R

```diff
@@ -0,0 +1,98 @@
+DataProcessor <- R6::R6Class(
+  classname = "lgb.DataProcessor",
+  public = list(
+    factor_levels = NULL,
+
+    process_label = function(label, objective, params) {
+
+      if (is.character(label)) {
+        label <- factor(label)
+      }
+
+      if (is.factor(label)) {
+
+        if (is.ordered(label)) {
+          warning("Passed labels as ordered factor. Order of levels is ignored.")
+        }
+
+        self$factor_levels <- levels(label)
+        if (length(self$factor_levels) <= 1L) {
+          stop("Labels to predict is a factor with <2 possible values.")
+        }
+
+        label <- as.numeric(label) - 1.0
+        out <- list(label = label)
+        if (length(self$factor_levels) == 2L) {
+          if (objective == "auto") {
+            objective <- "binary"
+          }
+          if (!(objective %in% .BINARY_OBJECTIVES())) {
+            stop("Two-level factors as labels only allowed for objective='binary' or objective='auto'.")
+          }
+        } else {
+          if (objective == "auto") {
+            objective <- "multiclass"
+          }
+          if (!(objective %in% .MULTICLASS_OBJECTIVES())) {
+            stop(
+              sprintf(
+                "Factors with >2 levels as labels only allowed for multi-class objectives. Got: %s (allowed: %s)"
+                , objective
+                , toString(.MULTICLASS_OBJECTIVES())
+              )
+            )
+          }
+          data_num_class <- length(self$factor_levels)
+          params <- lgb.check.wrapper_param(
+              main_param_name = "num_class"
+              , params = params
+              , alternative_kwarg_value = data_num_class
+          )
+          if (params[["num_class"]] != data_num_class) {
+            warning(
+              sprintf(
+                "Found num_class=%d in params, but 'label' is a factor with %d levels. 'num_class' will be ignored."
+                , params[["num_class"]]
+                , data_num_class
+              )
+            )
+            params$num_class <- data_num_class
```

Comment:
This behavior may make it difficult for users to resample models with outcomes that have class imbalances with this interface. I’m on board for a warning like that above, but I think this value ought to still respect `num_class` rather than the observed levels.

For example, with the following data:

``` r
set.seed(1)
y <- sample(c("small", "medium", "large"), size = 40, replace = TRUE, prob = c(.05, .9, .05))
y <- factor(y, levels = c("small", "medium", "large"))
x <- matrix(as.numeric(y) + rnorm(40))

y
#>  [1] medium medium medium large  medium medium large  medium medium medium
#> [11] medium medium medium medium medium medium medium small  medium medium
#> [21] large  medium medium medium medium medium medium medium medium medium
#> [31] medium medium medium medium medium medium medium medium medium medium
#> Levels: small medium large

head(x)
#>          [,1]
#> [1,] 2.918977
#> [2,] 2.782136
#> [3,] 2.074565
#> [4,] 1.010648
#> [5,] 2.619826
#> [6,] 1.943871
```

…I may take a bootstrap resample that doesn’t include one of `"small"` or
`"large"`.

``` r
set.seed(4)

indices <- sample(1:40, size = 40, replace = TRUE)
y_new <- y[indices]

y_new
#>  [1] medium medium large  medium medium medium medium medium medium medium
#> [11] medium medium medium medium medium medium medium medium medium medium
#> [21] medium large  medium medium medium medium medium medium medium medium
#> [31] medium medium medium medium medium medium medium medium medium medium
#> Levels: small medium large

unique(y_new)
#> [1] medium large 
#> Levels: small medium large
```

The `lgb.train()` interface handles this case fine:

``` r
lgb.train(
  data = lgb.Dataset(data = x[indices, , drop = FALSE], 
                     label = (as.numeric(y) - 1.0)[indices]),
  params = list(objective = "multiclass", num_class = 3),
  verbose = -1L
)
#> <lgb.Booster>
#>   Public:
#>     add_valid: function (data, name) 
#>     best_iter: -1
#>     best_score: NA
#>     current_iter: function () 
#>     dump_model: function (num_iteration = NULL, feature_importance_type = 0L) 
#>     eval: function (data, name, feval = NULL) 
#>     eval_train: function (feval = NULL) 
#>     eval_valid: function (feval = NULL) 
#>     finalize: function () 
#>     initialize: function (params = list(), train_set = NULL, modelfile = NULL, 
#>     lower_bound: function () 
#>     params: list
#>     predict: function (data, start_iteration = NULL, num_iteration = NULL, 
#>     raw: NA
#>     record_evals: list
#>     reset_parameter: function (params, ...) 
#>     rollback_one_iter: function () 
#>     save: function () 
#>     save_model: function (filename, num_iteration = NULL, feature_importance_type = 0L) 
#>     save_model_to_string: function (num_iteration = NULL, feature_importance_type = 0L) 
#>     set_train_data_name: function (name) 
#>     to_predictor: function () 
#>     update: function (train_set = NULL, fobj = NULL) 
#>     upper_bound: function () 
#>   Private:
#>     eval_names: NULL
#>     get_eval_info: function () 
#>     handle: lgb.Booster.handle
#>     higher_better_inner_eval: NULL
#>     init_predictor: NULL
#>     inner_eval: function (data_name, data_idx, feval = NULL) 
#>     inner_predict: function (idx) 
#>     is_predicted_cur_iter: list
#>     name_train_set: training
#>     name_valid_sets: list
#>     num_class: 3
#>     num_dataset: 1
#>     predict_buffer: list
#>     set_objective_to_none: FALSE
#>     train_set: lgb.Dataset, R6
#>     train_set_version: 1
#>     valid_sets: list
```

If the factor still has those levels, and if I set `num_class` to reflect that small and large are still possible outcomes, I’d expect that I’d still be able to train with this resample.

<sup>Created on 2023-02-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
If you subset a factor, it retains the levels, so that shouldn't be an issue.